File: cpp/basic_syntax/variables_loops_conditions.cpp
/**
 * @file variables_loops_conditions.cpp
 * @brief Demonstration of variables, loops, and conditional statements in C++
 *
 * This example shows:
 * - Various variable declarations and types including auto
 * - for, range-based for, while, and do-while loops
 * - if, else if, else conditions
 * - switch statements
 */

#include <iostream>
#include <vector>
#include <string>

/**
 * @brief Program entry point
 * @return Exit status code
 */
int main() {
    // Variable declarations
    int i = 0;
    float f = 3.14f;
    double d = 2.71828;
    char c = 'A';
    bool b = true;
    std::string s = "Hello, C++";
    auto a = 42;  // Type inference
    
    // Standard array
    int array[5] = {1, 2, 3, 4, 5};
    
    // STL container
    std::vector<int> vec = {10, 20, 30, 40, 50};
    
    // Standard for loop
    std::cout << "Standard for loop example:" << std::endl;
    for (i = 0; i < 5; i++) {
        std::cout << "array[" << i << "] = " << array[i] << std::endl;
    }
    
    // Range-based for loop (C++11)
    std::cout << "\nRange-based for loop example:" << std::endl;
    for (const auto& value : vec) {
        std::cout << "value: " << value << std::endl;
    }
    
    // While loop
    std::cout << "\nWhile loop example:" << std::endl;
    i = 0;
    while (i < 5) {
        std::cout << "iteration " << i << std::endl;
        i++;
    }
    
    // Do-while loop
    std::cout << "\nDo-while loop example:" << std::endl;
    i = 0;
    do {
        std::cout << "iteration " << i << std::endl;
        i++;
    } while (i < 5);
    
    // If-else statements
    std::cout << "\nIf-else example:" << std::endl;
    if (b) {
        std::cout << "b is true" << std::endl;
    } else {
        std::cout << "b is false" << std::endl;
    }
    
    // If-else if-else
    std::cout << "\nIf-else if-else example:" << std::endl;
    if (i < 3) {
        std::cout << "i is less than 3" << std::endl;
    } else if (i == 5) {
        std::cout << "i is equal to 5" << std::endl;
    } else {
        std::cout << "i is greater than 5" << std::endl;
    }
    
    // Switch statement
    std::cout << "\nSwitch example:" << std::endl;
    switch (c) {
        case 'A':
            std::cout << "c is 'A'" << std::endl;
            break;
        case 'B':
            std::cout << "c is 'B'" << std::endl;
            break;
        default:
            std::cout << "c is not 'A' or 'B'" << std::endl;
            break;
    }
    
    return 0;
}

File: python/basic_syntax/variables_loops_conditions.py
#!/usr/bin/env python3
"""
Demonstration of variables, loops, and conditional statements in Python

This example shows:
- Various variable declarations and types
- for and while loops
- if, elif, else conditions
- List, dictionary, and set comprehensions
- Type hints (Python 3.5+)
"""

from typing import List, Dict, Set, Tuple, Optional


def main() -> None:
    """
    Program entry point
    """
    # Variable declarations with type hints
    i: int = 0
    f: float = 3.14
    s: str = "Hello, Python"
    b: bool = True

    # Collection types
    my_list: List[int] = [1, 2, 3, 4, 5]
    my_dict: Dict[str, int] = {"one": 1, "two": 2, "three": 3}
    my_set: Set[int] = {10, 20, 30, 40, 50}
    my_tuple: Tuple[int, str, float] = (1, "two", 3.0)

    # For loop with a list
    print("For loop with list example:")
    for value in my_list:
        print(f"value: {value}")

    # For loop with range
    print("\nFor loop with range example:")
    for i in range(5):
        print(f"iteration: {i}")

    # For loop with enumeration
    print("\nFor loop with enumeration example:")
    for index, value in enumerate(my_list):
        print(f"index: {index}, value: {value}")

    # For loop with dictionary
    print("\nFor loop with dictionary example:")
    for key, value in my_dict.items():
        print(f"key: {key}, value: {value}")

    # While loop
    print("\nWhile loop example:")
    i = 0
    while i < 5:
        print(f"iteration: {i}")
        i += 1

    # If-else statements
    print("\nIf-else example:")
    if b:
        print("b is True")
    else:
        print("b is False")

    # If-elif-else
    print("\nIf-elif-else example:")
    if i < 3:
        print("i is less than 3")
    elif i == 5:
        print("i is equal to 5")
    else:
        print("i is greater than 5")

    # List comprehension
    print("\nList comprehension example:")
    squares = [x**2 for x in range(10) if x % 2 == 0]
    print(f"Squares of even numbers from 0-9: {squares}")

    # Dictionary comprehension
    print("\nDictionary comprehension example:")
    char_count = {char: s.count(char) for char in set(s) if char.isalpha()}
    print(f"Character count in '{s}': {char_count}")


if __name__ == "__main__":
    main()
